### PR TITLE
[WIP] Adding a geom package

### PR DESCRIPTION
**Pull request by luanpotter**
This is WIP!

I started developing a geom package for flame. It provides shapes and functions for geometric operations.

Since there is no extension functions in Dart, we don't take advantage of existing classes, like Rect: everything is re-created.

This classes use the Position class, but they are immutable. Position is not (for performance reasons).

I'm still deciding what is better. Maybe have two classes, because I could not figure out a way to make a class both mutable and immutable in Dart as well.

Probably would also like to add another breaking change to move Position inside the geom class, make rename it to Point.

For now this just adds the new files for review and ideas. Still figuring it out.

### CODE REVIEW

File: lib/geom/int_position.dart [hunk 1]
```diff
@@ -0,0 +1,60 @@
+import 'package:flame/position.dart';
```
Comment (erickzanardo) [inline lib/geom/int_position.dart]: Why do we need a Position class of Int? Is this just a copy of Position, but using integers?

File: lib/geom/rectangle.dart [hunk 2]
```diff
@@ -0,0 +1,211 @@
+import 'dart:math' as math;
+import 'dart:ui';
+
+import 'package:flame/position.dart';
```
Comment (erickzanardo) [inline lib/geom/rectangle.dart]: import 'position.dart'?

File: lib/geom/rectangle.dart [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  Rect asRect() {
+    if (isEmpty) {
+      return const Rect.fromLTWH(0.0, 0.0, 0.0, 0.0);
+    }
+    return Rect.fromLTWH(left, top, width, height);
+  }
+
+  Position toLT() {
```
Comment (erickzanardo) [inline lib/geom/rectangle.dart]: What does `toLT` means?

File: lib/geom/int_position.dart (hunk 1 above)
Comment (luanpotter) [reply lib/geom/int_position.dart]: yup, that is exactly it. for integer handling? hahaha using doubles in place of integers my lead to precision issues, I guess. also makes the apis nicer when using, avoiding casts. We use that a lot on tales, but I think often of getting rid of it. I don't have a strong opinion about it yet.

File: lib/geom/rectangle.dart (hunk 3 above)
Comment (luanpotter) [reply lib/geom/rectangle.dart]: LT=leftTop. that is the position at the left top of the rectangle. it's flutter Rect notation, will add a doc/comment thing

File: lib/geom/circle.dart [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+/// A circle is the locus of points in a 2D plane such as their distance to a fixed point ([center]) is less or equal than a fixed number ([radius]).
+/// The circumference is the 1D boundary of the circle.
+/// The circle can be expressed as the locus of points such as:
+/// (x - center.x)^2 + (y - center.y)^2 <= radius^2
+/// This class uses doubles to represent the coordinates. forrrest
+class Circle {
+  /// The center of this circle.
+  Position center;
+
+  int finalVallue;
```
Comment (luanpotter) [inline lib/geom/circle.dart:19]: oops!

**Comment by luanpotter** [comment]
Will migrate this to a `flame_geom` package.

**Comment by luanpotter** [comment]
I moved all this code to a new package [flame_geom](https://github.com/flame-engine/flame_geom), following in line with extract things to separated modules in flame.

I will close this issue for now, please head to the new repo for more details.

**Pull request closed without merging**